common/http_test.cpp

```diff
@@ -1395,6 +1395,181 @@ TEST(HttpTest, DestroyClientBeforeRequestComplete) {
 	EXPECT_FALSE(client_hit_body);
 }
 
+TEST(HttpTest, TestResponseBodyReader) {
+	TestEventLoop loop;
+
+	http::ServerConfig server_config;
+	http::Server server(server_config, loop);
+	vector<uint8_t> received_body;
+	server.AsyncServeUrl(
+		"http://127.0.0.1:" TEST_PORT,
+		[](http::ExpectedIncomingRequestPtr exp_req) {
+			ASSERT_TRUE(exp_req) << exp_req.error().String();
+		},
+		[](http::ExpectedIncomingRequestPtr exp_req) {
+			ASSERT_TRUE(exp_req) << exp_req.error().String();
+
+			auto result = exp_req.value()->MakeResponse();
+			ASSERT_TRUE(result);
+			auto resp = result.value();
+
+			resp->SetHeader("Content-Length", to_string(BodyOfXes::TARGET_BODY_SIZE));
+			resp->SetBodyReader(make_shared<BodyOfXes>());
+			resp->SetStatusCodeAndMessage(200, "Success");
+			resp->AsyncReply([](error::Error err) { ASSERT_EQ(error::NoError, err); });
+		});
+
+	http::ClientConfig client_config;
+	http::Client client(client_config, loop);
+	auto req = make_shared<http::OutgoingRequest>();
+	req->SetMethod(http::Method::GET);
+	req->SetAddress("http://127.0.0.1:" TEST_PORT);
+	vector<uint8_t> buf;
+	// Use a weird buf size, just to iron out more corner cases.
+	buf.resize(1235);
+	client.AsyncCall(
+		req,
+		[&received_body, &buf](http::ExpectedIncomingResponsePtr exp_resp) {
+			ASSERT_TRUE(exp_resp) << exp_resp.error().String();
+			auto resp = exp_resp.value();
+
+			auto content_length = resp->GetHeader("Content-Length");
+			ASSERT_TRUE(content_length);
+			ASSERT_EQ(content_length.value(), to_string(BodyOfXes::TARGET_BODY_SIZE));
+
+			auto body_writer = make_shared<io::ByteWriter>(received_body);
+			body_writer->SetUnlimited(true);
+			auto reader = resp->MakeBodyAsyncReader();
+			// Lambdas can't be recursive, so we need to use an explicit function object
+			// to refer to ourselves.
```

Comment:
You mean as a sort of `AsyncRepeatingReader`, which instead of calling the callback once, calls it until the stream is exhausted? That's a neat idea, and I think it should work, but only for readers, not writers, since only readers have a predictable termination condition (`err != error::NoError || read == 0`).

I don't want to spend too much time on it, but I can see how hard it is to make a convenience interface for it.

Reply:
Ah, no wait, the bool. You're right, that can be the termination condition.

Reply:
@vpodzime: This did indeed help. Check out the last commit. I didn't add a corresponding one for Writer, mostly because then I'd have to test it, but it can easily be done later in the same manner.

Rest of the smaller fixes were squashed into existing commits.